src/scheduler.ts

```diff
@@ -0,0 +1,144 @@
+import type { PiscinaWorker, PiscinaTask, RunTaskOptions } from '.';
+
+function isTaskSchedulerLike (obj: {}): obj is TaskScheduler {
+  if (Object.getPrototypeOf(obj) === TaskScheduler) return true;
+
+  const keys: (keyof TaskScheduler)[] = [
+    'onWorkerAvailable',
+    'onAvailable',
+    'pick',
+    'delete',
+    'size'
+  ];
+
+  for (const key of keys) {
+    if (!(key in obj)) return false;
+  }
+
+  return true;
+}
+
+class TaskScheduler {
+  size: number;
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  constructor (_concurrentTasksPerWorker: number) {
+    this.size = 0;
+  }
+
+  [Symbol.iterator] (): IterableIterator<PiscinaWorker> {
+    throw new Error('Iterator not implemented.');
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  add (_worker: PiscinaWorker): void {
+    throw new Error('add Method not implemented.');
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  pick (_task: PiscinaTask, _runOptions: RunTaskOptions): PiscinaWorker | null {
+    throw new Error('pick Method not implemented.');
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  onAvailable (_cb: (worker: PiscinaWorker) => void): void {
+    throw new Error('onAvailable Method not implemented.');
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  delete (_worker: PiscinaWorker): void {
+    throw new Error('delete Method not implemented.');
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  onWorkerAvailable (_worker: PiscinaWorker): void {
+    throw new Error('onNewWorker Method not implemented.');
+  }
+
+  getAvailableCapacity (): number {
+    throw new Error('getAvailableCapacity Method not implemented.');
+  }
+}
+
+class DefaultTaskScheduler extends TaskScheduler {
+  #pendingItems = new Set<PiscinaWorker>();
+  #readyItems = new Set<PiscinaWorker>();
+  #maximumUsage: number;
+  #onAvailableListeners: ((item: PiscinaWorker) => void)[];
+
+  constructor (maximumUsage: number) {
```

Comment:
I think we could default it to `os.availableParallelism()` https://nodejs.org/api/os.html#osavailableparallelism

Reply:
Let me make an issue out of that 👍